### ISSUE TITLE
Different extension loading behaviour between Rubocop and Standard

### ISSUE THREAD

**Issue by kpumuk** [testdouble/standard#507]
With the [amazing feature](https://github.com/testdouble/standard/pull/506) that just landed in Standard, we have got an ability to load and configure custom cops and extensions. The behaviour is slightly differs from Rubocop.

## Reproducing the problem

Looking at [rubocop-rspec](https://github.com/rubocop/rubocop-rspec) as an example. In vanilla Rubocop, we do add this to `.rubocop.yml`:

```yaml
require:
  - rubocop-rspec
```

This loads the gem and the default configuration automatically.

Doing the same Standard:

`.standard.yml`
```yaml
extend_config:
  - .rubocop-rspec.yml
```

`.rubocop-rspec.yml`
```yaml
require:
  - rubocop-rspec
```

This loads the gem, but not the config, and now an exception will be thrown when a cop is run:

```
Scanning <home>/github/meta-tags/spec/configuration_spec.rb
undefined method `[]' for nil:NilClass

          RuboCop::RSpec::Language.config = config['RSpec']['Language']
                                                           ^^^^^^^^^^^^
<gems>/rubocop-rspec-2.17.0/lib/rubocop/cop/rspec/base.rb:22:in `on_new_investigation'
```

## Workaround

To load configuration we would need to add `inherit_gem` (more examples can be found in https://github.com/kpumuk/meta-tags/pull/254):

`.rubocop-rspec.yml`
```yaml
require:
  - rubocop-rspec
inherit_gem:
  rubocop-rspec:
    - config/default.yml
```

I think this workaround should either be documented, or the behaviour needs to be synced between Rubocop and Standard.

**Comment by searls** [comment]
I did a little research into this, and I assume it has something to do with path resolution by the [FeatureLoader](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/feature_loader.rb) when the [ConfigLoader](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/config_loader.rb) calls it? Since this feature invokes the `ConfigLoader` only to extract cop rules (and relies on the fact that the require cache is primed as a side effect), I suspect we're not doing enough to ensure that paths are required / load paths are correct. Maybe @koic can confirm that this sounds about right?

I may not have time to tackle this right away, but if you want to dive in and try to diagnose I'd appreciate it!

**Comment by searls** [comment]
I really puzzled over this all morning trying to figure out how the `ConfigLoader` was doing this and the answer is: it isn't. Gems that work by simply requiring them seem to all execute a monkey-patch at load-time.

For example, [rubocop-rails.rb:13](https://github.com/rubocop/rubocop-rails/blob/master/lib/rubocop-rails.rb#L13) invokes [Inject.defaults](https://github.com/rubocop/rubocop-rails/blob/5425d8ac6ad714bbdc4968fe580704ba31a729f0/lib/rubocop/rails/inject.rb#L8) which merges its own `config/default.yml` with the default (see: `ConfigLoader.merge_with_default(config, path, unset_nil: false)`)

You can see the same thing happening here in [rubocop-rspec](https://github.com/rubocop/rubocop-rspec/blob/master/lib/rubocop-rspec.rb#L42) and its [identical `Inject.defaults!` method](https://github.com/rubocop/rubocop-rspec/blob/68252e8311ddc8c83f41c3571f30b5eff6b0b7ac/lib/rubocop/rspec/inject.rb#L8).

I'm unsure how I want to handle this to be honest.

**Comment by kpumuk** [comment]
Maybe, simply updating the README to include the workaround would suffice? The "injection" process sounds fishy, and extracting the loaded configuration for an extension from Rubocop defaults seems very unreasonable (or even impossible).

**Comment by searls** [comment]
I _could_ load and copy the RuboCop defaults, load (or re-load) a given gem, then do a deep diff of the updated hash (and then re-restore the RuboCop defaults for the next extension gem), but that's a lot of extra cycles to spend on an activity that (given RuboCop doesn't currently have a proper plugin system) is probably better solved with more explicit configuration (I did update the README with a note on this earlier today)

**Comment by palkan** [comment]
Just stumbled across this one, too.

> solved with more explicit configuration

That will work for purely _configurational_ extensions. However, some RuboCop plugins also _inject_ custom login into other life cycle processes (e.g., [rubocop-md](https://github.com/rubocop/rubocop-md)).

I will take a deeper look at how Standard's extensions work; maybe, there is a way to make extensions work interchangeably with RuboCop.

**Comment by palkan** [comment]
> some RuboCop plugins also inject custom login into other life cycle processes

Nvm, that should work, I think, since we call `#load_file` (and thus, _require_ everything).

> RuboCop doesn't currently have a proper plugin system

We can consider the `inject!` approach used by `rubocop-rspec` to be a semi-official one 🙂 (since it's used by all extensions know to me). And I have an idea how to deal with it, will try soon.

**Comment by koic** [comment]
I missed this one. The same issue was recently opened in RuboCop:
https://github.com/rubocop/rubocop/issues/11451

@palkan Yeah, I would be very grateful if you could take this issue :-)

**Comment by searls** [comment]
#512 landed in 1.22.1 to fix this
